pkg/cache/stats_counter.go

```diff
@@ -0,0 +1,216 @@
+func (c *burrowStatsCounter) RecordMisses(miss uint64) {
+	c.miss.Add(float64(miss))
+}
+
+// RecordLoadSuccess records the number of successful loads.
+// This method is called by Get methods on a successful load.
+func (c *burrowStatsCounter) RecordLoadSuccess(loadTime time.Duration) {
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	c.loadSuccess.Inc()
+	c.loadTotalTime.Observe(loadTime.Seconds())
+}
+
+// RecordLoadError records the number of failed loads.
+// This method is called by Get methods on a failed load.
+func (c *burrowStatsCounter) RecordLoadError(loadTime time.Duration) {
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	c.loadError.Inc()
+	c.loadTotalTime.Observe(loadTime.Seconds())
+}
+
+// RecordEviction records the number of evictions.
+func (c *burrowStatsCounter) RecordEviction() {
+	c.eviction.Inc()
+}
+
+// Snapshot records the current stats.
+// This method is called only by Stats method.
+func (c *burrowStatsCounter) Snapshot(s *burrow.Stats) {
+	var err error
+	s.HitCount, err = collectCounter(c.hits)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache hits", "err", err)
+	}
+	s.MissCount, err = collectCounter(c.miss)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache misses", "err", err)
+	}
+	s.EvictionCount, err = collectCounter(c.eviction)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache evictions", "err", err)
+	}
+
+	if !c.trackLoadingCacheStats {
+		return
+	}
+	s.LoadSuccessCount, err = collectCounter(c.loadSuccess)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache load success", "err", err)
+	}
+	s.LoadErrorCount, err = collectCounter(c.loadError)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache load error", "err", err)
+	}
+	totalTime, err := collectHistogramSum(c.loadTotalTime)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to collect cache load time", "err", err)
+	}
+	s.TotalLoadTime = time.Duration(totalTime)
+}
+
+func collect(col prometheus.Collector) (*dto.Metric, error) {
+	// This could be done with prometheus.CounterFunc and atomic.Counter easier,
+	// but it would make it harder to use the same code for histograms.
+
+	// Snapshot function is called manually and we don't plan to use it.
+	// For completeness, we implement it.
+	var (
+		m     prometheus.Metric
+		count int
+		ch    = make(chan prometheus.Metric)
+		done  = make(chan struct{})
+	)
+
+	go func() {
+		for m = range ch {
+			count++
+		}
+		close(done)
+	}()
+
+	col.Collect(ch)
+	close(ch)
+	<-done
+
+	if count != 1 {
+		return nil, fmt.Errorf("collected %d metrics instead of exactly 1", count)
+	}
+
+	pb := &dto.Metric{}
+	if err := m.Write(pb); err != nil {
+		return nil, fmt.Errorf("error happened while collecting metrics: %w", err)
+	}
+
+	return pb, nil
+}
+
+func collectCounter(col prometheus.Collector) (uint64, error) {
```

Comment:
I think `collect` is a widely established term in Prometheus world, I think we should use a different term here, even `getCurrentCounterValue` would be better as it's super unclear what this does and why.